electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +264,136 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+    # Pfet
+    self.fet = self.Block(Fet.PFet(
+      drain_voltage=(0, self.pwr_out.link().voltage.upper()),
+      drain_current=output_current_draw,
+      # gate voltage accounts for a possible power on transient
+      gate_voltage=(- self.pwr_out.link().voltage.upper(),  self.pwr_out.link().voltage.upper()),
+    ))
+
+    self.res = self.Block(Resistor(self.gate_resistor))
+    # TODO: generate zener diode for high voltage applications
+    #  self.diode = self.Block(ZenerDiode(self.clamp_voltage))
+
+    self.import_kicad(
+      self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
+      conversions={
+        'pwr_in': VoltageSink(
+          current_draw=output_current_draw,
+        ),
+        'pwr_out': VoltageSource(
+          voltage_out=self.pwr_in.link().voltage,
+         ),
+        'gnd': Ground(),
+      })
+
+
+
+class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+  class PmosChargerReverseProtection(PowerConditioner, KiCadSchematicBlock, Block):
+    """PMOS Charger Reverse Protection.
+
+    PMOS transistors are used in charging circuits to prevent reverse voltage from damaging the battery or the circuit.
+    They act as a switch that allows current flow only in the correct direction. Regular PMOS reverse protection may not be suitable for bidirectional circuit.
+
+    Ports:
+        - pwr_in: Power in from the battery
+        - pwr_out: Power out for the load
+        - chg_in: Charger input to charge the battery
+        - chg_out: Charging output for the battery
+        - gnd: Ground
+
+    Note:
+        - Limitation: The highest Vgs/Vds are bounded by the transistors' specifications.
+        - There is also a rare case when this circuit being disconnected when a charger is connected first. But always reverse protect.
+        - R1 and R2 are the pullup bias resistors for mp1 and mp2 PFet.
+
+    More info at: https://www.edn.com/reverse-voltage-protection-for-battery-chargers/
+    """
+
+  @init_in_parent
+  def __init__(self, r1_val: RangeLike = 100*kOhm(tol=0.01), r2_val: RangeLike = 100*kOhm(tol=0.01), rds_on: RangeLike =(0, 0.1)*Ohm):
+    super().__init__()
+
+    self.gnd = self.Port(Ground.empty(), [Common])
+    self.pwr_out = self.Port(VoltageSource.empty(),)  # Load
+    self.pwr_in = self.Port(VoltageSink.empty(), )  # For connecting battery pwr output
+    self.chg_in = self.Port(VoltageSink.empty(), )  # Charger input
+    self.chg_out = self.Port(VoltageSource.empty(),)  # Charging output to lipo
+
+    self.r1_val = self.ArgParameter(r1_val)
+    self.r2_val = self.ArgParameter(r2_val)
+    self.rds_on = self.ArgParameter(rds_on)
+
+  def contents(self):
+    super().contents()
+    self.r1 = self.Block(Resistor(resistance=self.r1_val))
+    self.r2 = self.Block(Resistor(resistance=self.r2_val))
+
+    batt_voltage = self.pwr_in.link().voltage.hull(self.chg_out.link().voltage).hull((0, 0))
```

Comment:
```suggestion
    batt_voltage = self.pwr_in.link().voltage.hull(self.chg_in.link().voltage).hull(self.gnd.link().voltage)
```
a bit more principled, using the ground link and consistently using source voltages